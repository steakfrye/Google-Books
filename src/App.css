.pad {
  margin-top: 10%;
}
