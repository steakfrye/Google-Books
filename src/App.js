import React, { Component } from 'react';
import { BrowserRouter as Router, Route } from 'react-router-dom';

import './App.css';
import LandingPage from './components/LandingPage';
import Table from './components/Table';
import Navbar from './components/Navbar';
import NoResults from './components/NoResults';

const key = require('./config/keys');

class App extends Component {
  constructor() {
    super();
    this.state = {
      searchResults: [],
      terms: '',
      start: 0,
      redirect: false,
      error: '',
    };
    this.onSearch = this.onSearch.bind(this);
  }

  // Handles search errors and fetches data
  onSearch(terms, start) {
    let filteredTerms = terms.replace(/\s/g, '+');

    if (!filteredTerms || filteredTerms.length === 0) {
      this.setState({ searchResults: [], error: 'You must submit a valid search.', redirect: false });
    } else {
      fetch(`/volumes?q=${filteredTerms}&startIndex=${start}&key=${key.secretKey}`)
        .then(res => {
            if (!res.ok) {
              throw Error(res.statusText);
            }

            return res.json();
          })
        .then(data => {
          // Store data and redirect
          console.log(data);
          this.setState({ terms: terms, start: start, searchResults: data.items, redirect: true, error: '' });
        })
        .catch(err => {
          console.log(err);
          this.setState({ searchResults: [], error: 'You must submit a valid search.', redirect: false });
        });
    }
  }

  render() {
    const { searchResults, terms, start, redirect, error } = this.state;

    return (
      <Router>
        <div className="app">
          <Route exact path="/" component={() =>
            <LandingPage
              onSearch={this.onSearch}
              error={error}
              redirect={redirect}
            />
          }/>
          <Route path="/search" component={() =>
            <div>
              <Navbar onSearch={this.onSearch} error={error} />
              {
                (!searchResults || searchResults.length === 0) ?
                  <NoResults /> :
                <Table
                  books={searchResults}
                  onSearch={this.onSearch}
                  terms={terms}
                  start={start}
                />
              }
            </div>
          }/>
        </div>
      </Router>
    );
  }
}

export default App;
