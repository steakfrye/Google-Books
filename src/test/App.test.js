import React from 'react';
import ReactDOM from 'react-dom';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import App from '../App';

configure({ adapter: new Adapter() });

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('<App />', () => {
  const app = mount(<App />);

  it('displays landing page', () => {
    const tree = renderer
      .create(app)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('does not create error on valid params', () => {
    const instance = app.instance();
    expect(app.state('searchResults')).toHaveLength(0);
    instance.onSearch('dog', 0);
    expect(app.state('error')).toBe('');
  });

  it('creates error on invalid params', () => {
    const instance = app.instance();
    instance.onSearch('', 0);
    expect(app.state('error')).toBe('You must submit a valid search.');
    instance.onSearch('   ', 0);
    expect(app.state('error')).toBe('You must submit a valid search.');
  });
});
