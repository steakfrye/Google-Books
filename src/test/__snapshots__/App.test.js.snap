// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> displays landing page 1`] = `
<div
  className="app"
>
  <div
    className="container pad"
  >
    <div
      className="row align-items-center"
    >
      <div
        className="col-md-12 text-center"
      >
        <h1
          className="display-3 mb-4"
        >
          Book Search
        </h1>
        <hr />
        <h3>
          Use the search to find a list of relevant books.
        </h3>
        <div>
          <form
            className="form-group row justify-content-center"
            onSubmit={[Function]}
          >
            <input
              aria-label="Search"
              className="form-control col-sm-4"
              onChange={[Function]}
              placeholder="Search Books"
              type="search"
              value=""
            />
            <button
              className="btn btn-outline-success my-2 my-sm-0"
              type="submit"
            >
              Search
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
